tests/async_tests/test_connection_pool.py

```diff
@@ -53,14 +57,59 @@ def is_connection_dropped(self) -> bool:
         return False
 
 
+class AlwaysPendingConnection(MockConnection):
+    async def arequest(
+        self,
+        method: bytes,
+        url: URL,
+        headers: Headers = None,
+        stream: httpcore.AsyncByteStream = None,
+        ext: dict = None,
+    ) -> Tuple[int, Headers, httpcore.AsyncByteStream, dict]:
+        result = await super().arequest(method, url, headers, stream, ext)
+        self.state = ConnectionState.PENDING
+        return result
+
+
+class RefusedConnection(MockConnection):
+    async def arequest(
+        self,
+        method: bytes,
+        url: URL,
+        headers: Headers = None,
+        stream: httpcore.AsyncByteStream = None,
+        ext: dict = None,
+    ) -> Tuple[int, Headers, httpcore.AsyncByteStream, dict]:
+        raise ConnectionRefusedError
+
+
 class ConnectionPool(httpcore.AsyncConnectionPool):
-    def __init__(self, http_version: str):
+    def __init__(
+        self,
+        http_version: str,
+        connection_class: Type = MockConnection,
+    ):
         super().__init__()
         self.http_version = http_version
+        if http_version == "HTTP/2":
+            self._http2 = True
+        self.connection_class = connection_class
         assert http_version in ("HTTP/1.1", "HTTP/2")
 
     def _create_connection(self, **kwargs):
-        return MockConnection(self.http_version)
+        return self.connection_class(self.http_version)
+
+
+class RequiringNewConnectionPool(httpcore.AsyncConnectionPool):
```

Comment:
Connection pool tries to reuse the connection acquired from the pool. But having tried to reuse closed connection, it receives `NewConnectionRequired` error